tests/testsuite/resolve.rs

```diff
@@ -185,6 +190,97 @@ fn loc_names(names: &[(&'static str, &'static str)]) -> Vec<PackageId> {
         .collect()
 }
 
+/// This generates a random registry index.
+/// Unlike vec((Name, Ver, vec((Name, VerRq), ..), ..)
+/// This strategy has a high probability of having valid dependencies
+fn registry_strategy() -> impl Strategy<Value=Vec<Summary>> {
+    const VALID_NAME_STRATEGY: &str = "[A-Za-z_-][A-Za-z0-9_-]*";
+    const MAX_CRATES: usize = 10;
+    const MAX_VERSIONS: usize = 10;
+
+    fn range(max: usize) -> impl Strategy<Value = (usize, usize)> {
+        (0..max).prop_flat_map(move |low| (Just(low), low..=max))
+    }
+
+    fn ver(max: usize) -> impl Strategy<Value=(usize, usize, usize)> {
+        (0..=max, 0..=max, 0..=max)
+    }
+
+    btree_map(
+        VALID_NAME_STRATEGY,
+        btree_set(
+            ver(MAX_VERSIONS).prop_map(|(a, b, c)| format!("{}.{}.{}", a, b, c)),
+            1..=MAX_VERSIONS,
+        ),
+        1..=MAX_CRATES,
+    ).prop_map(|mut b| {
+        // root is the name of the thing being compiled
+        // so it would be confusing to have it in the index
+        b.remove("root");
+        // if randomly pointing to a dependency that does not exist then call it `bad`
+        b.insert("bad".to_owned(), BTreeSet::new());
+        (
+            b.iter()
+                .map(|(name, _)| name.clone())
+                .collect::<Vec<String>>(),
+            b.iter()
+                .flat_map(|(name, vers)| vers.into_iter().map(move |v| (name.clone(), v.clone())))
+                .collect::<Vec<_>>(),
+        )
+    }).prop_flat_map(|(names, data)| {
+        let names_len = names.len();
+        let data_len = data.len();
+        (
+            Just(data),
+            vec(subsequence(names, 0..names_len), data_len..=data_len),
+        )
+    }).prop_flat_map(|(data, deps)| {
+        let len = deps.iter().map(|x| x.len()).sum();
+        (
+            Just(data),
+            Just(deps),
+            vec(
```

Comment:
I think you could lift this out to a top-level strategy by just generating a vector of size `MAX_CRATES*MAX_CRATES` since the code below only looks at the elements it needs. That would remove another layer of flat mapping.